File: cpp/Linked Lists/linked_list_intersection.cpp
#include "ds/ListNode.h"
using ds::ListNode;

/**
 * Definition of ListNode:
 * struct ListNode {
 *     int val;
 *     ListNode* next;
 *     ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
 * };
 */

ListNode* linkedListIntersection(ListNode* head_A, ListNode* head_B) {
    ListNode* ptr_A = head_A;
    ListNode* ptr_B = head_B;
    // Traverse through list A with 'ptr_A' and list B with 'ptr_B' 
    // until they meet.
    while (ptr_A != ptr_B) {
        // Traverse list A -> list B by first traversing 'ptr_A' and 
        // then, upon reaching the end of list A, continue the 
        // traversal from the head of list B.
        if (ptr_A != nullptr) {
            ptr_A = ptr_A->next;
        } else {
            ptr_A = head_B;
        }
        // Simultaneously, traverse list B -> list A.
        if (ptr_B != nullptr) {
            ptr_B = ptr_B->next;
        } else {
            ptr_B = head_A;
        }
    }
    // At this point, 'ptr_A' and 'ptr_B' either point to the 
    // intersection node or both are null if the lists do not 
    // intersect. Return either pointer.
    return ptr_A;  
}

File: csharp/Linked Lists/LinkedListIntersection.cs
using ds.ListNode;

/*
	Definition of ListNode:
	class ListNode
	{
		public ListNode(int val, ListNode next = null)
		{
			Val = val;
			Next = next;
		}

		public int Val { get; set; }

		public ListNode Next { get; set; }
	}
*/

public class Solution
{
	public ListNode LinkedListIntersection(ListNode headA, ListNode headB)
	{
		ListNode ptrA = headA;
		ListNode ptrB = headB;

		// Traverse through list A with 'ptrA' and list B with 'ptrB' until they meet.
		while (ptrA != ptrB)
		{
			// Traverse list A -> list B by first traversing 'ptrA' and
			// then, upon reaching the end of list A, continue the 
			// traversal from the head of list B.
			ptrA = ptrA != null ? ptrA.Next : headB;

			// Simultanously, traverse list B -> list A.
			ptrB = ptrB != null ? ptrB.Next : headA;
		}

		// At this point, 'ptrA' and 'ptrB' either point to the 
		// intersection node or both are null if the lists do not 
		// intersect. Return either pointer.
		return ptrA;
	}
}
